File: linearalgebra.py
import math

class Vector3():
    def __init__(self, x=1.0, y=1.0, z=1.0):
        self.x , self.y, self.z = x, y, z

    def magnitude(self):
        return math.sqrt(self.x**2 + self.y**2 + self.z**2)

    def normalized(self):
        return self / self.magnitude()

    def __str__(self):
        return "{0} {1} {2}".format(self.x, self.y, self.z)

    # USES % FOR CROSS PRODUCT
    def __mod__(self, other):
        return Vector3(x=(self.y * other.z - self.z * other.y),
                       y=-(self.x * other.z - self.z * other.x),
                       z=(self.x * other.y - self.y * other.x))

    # USES ** FOR DOT PRODUCT
    def __pow__(self, other):
        return self.x * other.x + self.y * other.y + self.z * other.z

    def __add__(self, other):
        if isinstance(other, Vector3):
            return Vector3(x=self.x + other.x,
                           y=self.y + other.y,
                           z=self.z + other.z)
        else:
            return Vector3(x=self.x + other,
                           y=self.y + other,
                           z=self.z + other)

    def __sub__(self, other):
        if isinstance(other, Vector3):
            return Vector3(x=self.x - other.x,
                           y=self.y - other.y,
                           z=self.z - other.z)
        else:
            return Vector3(x=self.x - other,
                           y=self.y - other,
                           z=self.z - other)


    def __mul__(self, other):
        if isinstance(other, Vector3):
            return Vector3(x=self.x * other.x,
                           y=self.y * other.y,
                           z=self.z * other.z)
        else:
            return Vector3(x=self.x * other,
                           y=self.y * other,
                           z=self.z * other)

    def __truediv__(self, scalar):
        return Vector3(x=self.x / scalar,
                       y=self.y / scalar,
                       z=self.z / scalar)

    def __div__(self, scalar):
        return Vector3(x=self.x / scalar,
                       y=self.y / scalar,
                       z=self.z / scalar)
    __rmod__ = __mod__
    __rpow__ = __pow__
    __radd__ = __add__
    __rsub__ = __sub__
    __rmul__ = __mul__
    __rtruediv__ = __truediv__
    __rdiv__ = __div__

class Matrix3():
    def __init__(self,
                 m=[1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0],
                 rows=(None, None, None),
                 cols=(None, None, None)):
        self.m = m
        for i, row in enumerate(rows):
            if row is not None:
                self.m[i] = row.x
                self.m[i + 1] = row.y
                self.m[i + 2] = row.z

        for i, col in enumerate(cols):
            if col is not None:
                self.m[(3 * i)] = col.x
                self.m[(3 * i) + 1] = col.y
                self.m[(3 * i) + 2] = col.z

    def row(self, i):
        return Vector3(self.m[i], self.m[i + 1], self.m[i + 2])

    def col(self, i):
        return Vector3(self.m[(3 * i)], self.m[(3 * i) + 1], self.m[(3 * i) + 2])

    def __mul__(self, other):
        if isinstance(other, Matrix3):
            m = [
                self.row(0) ** other.col(0),
                self.row(0) ** other.col(1),
                self.row(0) ** other.col(2),
                self.row(1) ** other.col(0),
                self.row(1) ** other.col(1),
                self.row(1) ** other.col(2),
                self.row(2) ** other.col(0),
                self.row(2) ** other.col(1),
                self.row(2) ** other.col(2)
            ]
            return Matrix3(m=m)

        elif isinstance(other, Vector3):
            x = self.row(0) ** other
            y = self.row(1) ** other
            z = self.row(2) ** other
            return Vector3(x=x, y=y, z=z)
        else:
            return None

File: linearalgebra.cpp
#include "linearalgebra.h"
#include<stdio.h>
#include<cmath>

float Float3::magnitude() {
  return sqrt((*this)^(*this));
}
Float3 Float3::normalized() {
  return (*this) / magnitude();
}
Float3 Float3::operator +(Float3 other) {
  return Float3(x+other.x, y+other.y, z+other.z);
}
void Float3::operator +=(Float3 other) {
  x += other.x;
  y += other.y;
  z += other.z;
  return;
}
Float3 Float3::operator +(float scalar) {
  return Float3(x+scalar, y+scalar, z+scalar);
}
void Float3::operator +=(float scalar) {
  x += scalar;
  y += scalar;
  z += scalar;
  return;
}
Float3 Float3::operator -(Float3 other) {
  return Float3(x-other.x, y-other.y, z-other.z);
}
void Float3::operator -=(Float3 other) {
  x -= other.x;
  y -= other.y;
  z -= other.z;
  return;
}
Float3 Float3::operator -(float scalar) {
  return Float3(x+scalar, y+scalar, z+scalar);
}
void Float3::operator -=(float scalar) {
  x -= scalar;
  y -= scalar;
  z -= scalar;
  return;
}
Float3 Float3::operator *(Float3 other) {
  return Float3(x*other.x, y*other.y, z*other.z);
}
void Float3::operator *=(Float3 other) {
  x *= other.x;
  y *= other.y;
  z *= other.z;
  return;
}
Float3 Float3::operator *(float scalar) {
  return Float3(x*scalar, y*scalar, z*scalar);
}
void Float3::operator *=(float scalar) {
  x *= scalar;
  y *= scalar;
  z *= scalar;
  return;
}
Float3 Float3::operator %(Float3 other) {
  return Float3(y*other.z - z*other.y,
                x*other.z - z*other.x,
                x*other.y - y*other.x);
}
void Float3::operator %=(Float3 other) {
  x = y*other.z - z*other.y;
  y = x*other.z - z*other.x;
  z = x*other.y - y*other.x;
  return;
}
float Float3::operator ^(Float3 other) {
  return x * other.x + y * other.y + z * other.z;
}
float Float3::operator [](int index) {
  switch(index) {
    case 0:
      return x;
      break;
    case 1:
      return y;
      break;
    case 2:
      return z:
      break;
    default:
      return -1.0;
      break;
  }
}

Matrix3::Matrix3() {
      Float vectors[3] = {
        Float3(1.0, 0.0, 0.0), Float3(0.0, 1.0, 0.0), Float3(0.0, 0.0, 1.0)
      };
      _rows = vectors;
      _cols = vectors;
    }

Matrix3::Matrix3(Float3 vectors[3], bool rows) {
  if (rows) {
    _rows = vectors;
  } else {
    _cols = vectors;
  }
  for (int i = 0; i < 3; i++) {
    if (rows) {
      _cols[i] = Float3(vectors[0][i], vectors[1][i], vectors[2][i]);
    } else {
      _rows[i] = Float3(vectors[0][i], vectors[1][i], vectors[2][i]);
    }
  }
}

void Matrix3::get_col(int i) {
  return _cols[i];
}

void Matrix3::set_col(int i, Float3 vector) {
  _cols[i] = vector;
  for (int j = 0; j < 3; j++) {
    _rows[j][i] = vector[j];
  }
  return;
}

void Matrix3::get_row(int i) {
  return _rows[i];
}

void Matrix3::set_row(int i) {
  _rows[i] = vector;
  for (int j = 0; j < 3; j++) {
    _cols[j][i] = vector[j];
  }
  return;
}

void Matrix3::get_cell(int x, int y) {
  return _rows[x][y];
}

void Matrix3::set_cell(int x, int y, float value) {
  _rows[x][y] = value;
  _cols[y][x] = value;
}

Matrix3 Matrix3::operator *(Matrix3 other) {
  Matrix3 result = Matrix3();
  for (int i = 0; i < 3; i++) {
    for (int j = 0; j < 3; j++) {
      result.set_cell[i][j] = get_row(i)^get_col(j);
    }
  }
}

Float3 Matrix3::operator *(Float3 vector) {
  return Float3(get_row(0)^vector,
                get_row(1)^vector,
                get_row(2)^vector);
}
